### PR TITLE
Add support for annotating types with layouts.

### PR DESCRIPTION
**Pull request by goldfirere**
For example, this adds support for [(int : immediate)] and
[(string list : value)].

This is the first part of layout annotations. See the test case for a hint of what's to come! But this can still be merged as is, if we like.

### CODE REVIEW

**Comment by goldfirere** [comment]
This is now in an intermediate state as I'm adding more layout annotations. The first two commits are good, but please wait for general review. I will unmark this as a Draft and post here when it's ready for proper review.

**Comment by goldfirere** [comment]
This is now ready for review. There are tests failing in `testsuite/tests/typing-layouts/annots.ml`, but that's because I think there are type-inference bugs. Want to coordinate with @ccasin to fix those.

File: boot/menhir/menhirLib.ml [hunk 1]
```diff
@@ -1,14 +1,11 @@
 module General = struct
 (******************************************************************************)
 (*                                                                            *)
-(*                                   Menhir                                   *)
+(*                                    Menhir                                  *)
```
Comment (ccasin) [inline boot/menhir/menhirLib.ml]: It looks like you have Menhir version 20220210 on the switch you used to compile this.  For some reason that's not entirely clear to me, the compiler uses 20210419.  You should install that version (`opam install menhir.20210419`) and promote the parser again.  (You can see this change in version number at the top of `boot/menhir/parser.ml`, but I couldn't put this comment there due to limitations in the github interface).

File: parsing/ast_helper.ml [hunk 2]
```diff
@@ -511,11 +514,16 @@ module Type = struct
   let mk ?(loc = !default_loc) ?(attrs = [])
         ?(docs = empty_docs) ?(text = [])
       ?(params = [])
+      ?layout
       ?(cstrs = [])
       ?(kind = Ptype_abstract)
       ?(priv = Public)
       ?manifest
       name =
+    let layout_attrs = match layout with
```
Comment (ccasin) [inline parsing/ast_helper.ml]: This is `Option.to_list`, fwiw.

File: testsuite/tests/typing-unboxed-types/test.ml [hunk 3]
```diff
@@ -94,7 +94,7 @@ Error: This type cannot be unboxed because
 (* let rec must be rejected *)
 type t10 = A of t10 [@@ocaml.unboxed] [@@value];;
 [%%expect{|
-type t10 = A of t10 [@@value] [@@unboxed]
+type t10 = A of t10 [@@unboxed]
```
Comment (ccasin) [inline testsuite/tests/typing-unboxed-types/test.ml]: I don't love this change, and similar changes on other tests.  I think there's a `[@@value]` annotation here because it's needed (or at least that's the case for some of these), and dropping it from the output means what's printed won't have the same behavior as what was input.

While I agree we can mostly not print `[@@value]`, I think I'd prefer to bring back the behavior that we do print it if it was explicitly written in the source program.  (It would be even better to print it only if it's needed, but that seems hard.)

File: parsing/location.ml [hunk 4]
```diff
@@ -26,6 +26,12 @@ let in_file name =
 let none = in_file "_none_";;
 let is_none l = (l = none);;
 
+let combine { loc_start = start1; loc_end = _end1; loc_ghost = g1 }
+            { loc_start = _start2; loc_end = end2; loc_ghost = g2 } =
+  (* can't check ordering of positions, because there is no ordering
+     operator on positions *)
```
Comment (ccasin) [inline parsing/location.ml]: See [here](https://github.com/ocaml-flambda/ocaml-jst/blob/d78ff6de461fd42d809c314fee28d45793a82682/parsing/builtin_attributes.ml#L30) for one example of how to compare locations.  Or maybe that should just be factored out (and updated to include characters positions if the line positions are equal, which I've been meaning to do anyway).

Comment (ccasin) [reply parsing/location.ml]: Actually, scratch that - I don't think `combine` is used anywhere so it can probably just be deleted.

File: parsing/ast_mapper.ml [hunk 5]
```diff
@@ -85,6 +86,7 @@ let map_tuple3 f1 f2 f3 (x, y, z) = (f1 x, f2 y, f3 z)
 let map_opt f = function None -> None | Some x -> Some (f x)
 
 let map_loc sub {loc; txt} = {loc = sub.location sub loc; txt}
+let map_loc2 sub f {loc; txt} = {loc = sub.location sub loc; txt = f sub txt}
```
Comment (ccasin) [inline parsing/ast_mapper.ml]: Maybe `map_loc_txt`?  Or really any more descriptive name.

File: parsing/ast_mapper.mli [hunk 6]
```diff
@@ -81,6 +81,7 @@ type mapper = {
   include_declaration: mapper -> include_declaration -> include_declaration;
   include_description: mapper -> include_description -> include_description;
   label_declaration: mapper -> label_declaration -> label_declaration;
+  layout_annotation: mapper -> Asttypes.layout_annotation -> Asttypes.layout_annotation;
```
Comment (ccasin) [inline parsing/ast_mapper.mli]: If this proves tricky to keep when we back the parsetree changes out into attributes or the new extensions mechanism, I think it's fine to drop it.

File: parsing/ast_iterator.mli [hunk 7]
```diff
@@ -49,6 +49,7 @@ type iterator = {
   include_declaration: iterator -> include_declaration -> unit;
   include_description: iterator -> include_description -> unit;
   label_declaration: iterator -> label_declaration -> unit;
+  layout_annotation: iterator -> Asttypes.layout_annotation -> unit;
```
Comment (ccasin) [inline parsing/ast_iterator.mli]: Same comment as on `ast_mapper`.  This is nice to have, but fine to drop if it makes our lives easier when we do our "don't change the parsetree" shenanigans.

File: parsing/ast_helper.mli [hunk 8]
```diff
@@ -84,10 +84,13 @@ module Typ :
...
     val variant: ?loc:loc -> ?attrs:attrs -> row_field list -> closed_flag
                  -> label list option -> core_type
-    val poly: ?loc:loc -> ?attrs:attrs -> str list -> core_type -> core_type
+    val poly: ?loc:loc -> ?attrs:attrs -> str list -> core_type
+                 -> layout_annotation with_loc option list -> core_type
     val package: ?loc:loc -> ?attrs:attrs -> lid -> (lid * core_type) list
                  -> core_type
     val extension: ?loc:loc -> ?attrs:attrs -> extension -> core_type
+    val layout: ?loc:loc -> ?attrs:attrs -> core_type
+                 -> Asttypes.layout_annotation with_loc -> core_type
```
Comment (ccasin) [inline parsing/ast_helper.mli]: Can drop the `Asttypes` prefix.

File: parsing/ast_helper.mli [hunk 9]
```diff
@@ -236,7 +242,8 @@ module Te:
       str -> extension_constructor_kind -> extension_constructor
 
     val decl: ?loc:loc -> ?attrs:attrs -> ?docs:docs -> ?info:info ->
-      ?vars:str list -> ?args:constructor_arguments -> ?res:core_type ->
+      ?vars:(str list * layout_annotation with_loc option list) ->
```
Comment (ccasin) [inline parsing/ast_helper.mli]: I'd definitely prefer an additional layout parameter instead of making `vars` a pair of lists (or, even better, to make `vars` a list of pairs, both here and at all the relevant points in the parsetree, rather than having two lists and an invariant).  But I expect all this changes when we switch to use attributes or the new extensions mechanism, so it's not worth fixing this instant.

File: parsing/pprintast.mli [hunk 10]
```diff
@@ -53,3 +53,8 @@ val payload: Format.formatter -> Parsetree.payload -> unit
 val tyvar: Format.formatter -> string -> unit
   (** Print a type variable name, taking care of the special treatment
       required for the single quote character in second position. *)
+
+val layout_annotation: Format.formatter -> Asttypes.layout_annotation -> unit
+
+(* CR layouts RAE: where should this go? *)
+val layout_annotation_to_string : Asttypes.layout_annotation -> string
```
Comment (ccasin) [inline parsing/pprintast.mli]: It's tempting to do something`Type_layout.(to_string (of_layout_annotation l))`, instead of having a function here, so the two printers don't risk getting out of sync.  But this might be hard for dependency reasons?

File: parsing/printast.ml [hunk 11]
```diff
@@ -147,8 +147,20 @@ let arg_label i ppf = function
   | Labelled s -> line i ppf "Labelled \"%s\"\n" s
 ;;
 
-let typevars ppf vs =
-  List.iter (fun x -> fprintf ppf " %a" Pprintast.tyvar x.txt) vs
+let var_layout ?(print_quote = true) ppf (v, l) =
```
Comment (ccasin) [inline parsing/printast.ml]: I think default arguments are generally an anti-pattern and should be avoided, particularly for a function with just two call sites.  (It also seems a little odd to me to skip the `'` but not print it like a string in quotation marks).

File: parsing/pprintast.ml [hunk 12]
```diff
@@ -303,7 +310,18 @@ let tyvar ppf s =
   else
     Format.fprintf ppf "'%s" s
 
-let tyvar_loc f str = tyvar f str.txt
+let layout_annotation ppf lay =
+  Format.fprintf ppf "%s" (layout_annotation_to_string lay)
+
+let tyvar_layout_loc ?(print_quote = true) f (str,layout) =
```
Comment (ccasin) [inline parsing/pprintast.ml]: Same comment about default arguments as in `Printast`.

File: parsing/pprintast.ml [hunk 13]
```diff
@@ -1329,13 +1350,13 @@ and binding ctxt f {pvb_pat=p; pvb_expr=x; _} =
...
       | {ppat_desc=Ppat_constraint({ppat_desc=Ppat_var _} as pat,
-                                   {ptyp_desc=Ptyp_poly (args_tyvars, rt)});
+                                   {ptyp_desc=Ptyp_poly (args_tyvars, rt, _)});
          ppat_attributes=[]}->
           Some (pat, args_tyvars, rt)
       | _ -> None in
     let rec gadt_exp tyvars e =
       match e with
-      | {pexp_desc=Pexp_newtype (tyvar, e); pexp_attributes=[]} ->
+      | {pexp_desc=Pexp_newtype (tyvar, e, _); pexp_attributes=[]} ->
```
Comment (ccasin) [inline parsing/pprintast.ml:1362]: It's not obvious to me why the layouts are ignored here - it seems like somewhere we can have layout annotations, but maybe I'm misunderstanding what we're printing.

File: stdlib/camlinternalAtomic.mli [hunk 14]
```diff
@@ -20,8 +20,9 @@
    Stdlib__ prefix trick. *)
 
 type !'a t
-val make : 'a -> 'a t
+
 val get : 'a t -> 'a
+val make : 'a -> 'a t
 val set : 'a t -> 'a -> unit
```
Comment (ccasin) [inline stdlib/camlinternalAtomic.mli]: Accidental change?  This doesn't look related.  (But I'm not surprised you were playing around in this file, which is the absolute worst.)

File: parsing/parser.mly [hunk 15]
```diff
@@ -698,13 +698,85 @@ let mk_directive ~loc name arg =
...
+   Is that the beginning of, say, a function type where ['a] is the argument?
+   Or is it the beginning of a list of universal variables. To avoid conflicts
+   in this case (the non-terminal of interest is [possibly_poly], which can't
+   know which side to take), we have layout-annotation case of [typevar] parse
+   liberally, lining up with the layout-annotation case of [non_ident_atomic_type],
+   checking that the variable is really a variable and the layout annotation
+   is really a layout annotation in [check_type_var_from_type] and
+   [check_layout_from_type].
+*)
+
```
Comment (ccasin) [inline parsing/parser.mly:776]: This all seems hard, and there aren't a lot of good choices available.  Here's an annoying example:

```
type value
type x

module type Foo = sig
  val f : x:value -> int
  val g : (x:value) -> int
end
```

Here, `f` and `g` have very different types!  Now, it's already the case that parens are significant in function types, so this isn't necessarily catastrophic, but I do think it's already a source of confusion and we're just making that worse.

I'm wondering if we should make keywords for the layouts (e.g., `value_`), similar to the way local is handled.  I haven't worked out if that would simplify the ambiguity story above at all, but it might be a win regardless just for clarity.

File: parsing/parser.mly [hunk 16]
```diff
@@ -3216,14 +3310,25 @@ type_parameters:
...
+layout: (* : (layout_annotation * string) with_loc *)
+  (* this includes the string form of the layout annotation; useful for
+     some clients *)
+  ident { check_layout (make_loc $sloc) $1 }
+;
+
+layout_attr:
+  COLON
+  layout=layout
+    { Attr.mk ~loc:layout.loc (Location.map snd layout) (PStr []) }
```
Comment (ccasin) [inline parsing/parser.mly:3325]: It seems awkward to sometimes parse layouts as attributes and other times with their own parser.  But I assume this is destined for change when we adapt to the new extensions mechanism, and it just didn't make sense to substantially rework the existing layout annotations before that.

File: typing/printtyp.ml [hunk 17]
```diff
@@ -28,6 +28,63 @@ open Outcometree
...
+
+   We should print this layout annotation whenever it is not the
+   default [value]. (* CR layouts reisenberg: update when the default changes *)
+   This is a challenge, though, because the type in a [val] does not
+   explicitly quantify its free variables. So we must collect the free
+   variables, look to see whether any have interesting layouts, and
+   print the whole set of variables if any of them do. This is all
+   implemented in [extract_qtvs], used also in a number of other places
+   we do quantification (e.g. gadt-syntax constructors).
+*)
```
Comment (ccasin) [inline typing/printtyp.ml:87]: We've discussed this, but after looking through the examples I'm concerned that C3, in particular, is going to be very verbose in the fairly common case of code that is explicitly manipulating immediates.  

I think there is no way to do better without taking an environment.  The first thing is to look at the call sites of these functions and figure out if we even have access to the right environments.  Then we should talk more with Leo about whether using environments here is a real problem or just feels icky - I think maybe he's convinceable with examples. 

Even with an environment, we'd need to decide what to do.  How about: for variables with non-value layouts, we don't print the layout if the variable is used as an argument to a type constructor that requires the relevant parameter to have the exact layout on our variable.

File: typing/ctype.mli [hunk 18]
```diff
@@ -429,7 +429,11 @@ val nongen_class_declaration: class_declaration -> bool
            Uses the empty environment.  *)
 
 val free_variables: ?env:Env.t -> type_expr -> type_expr list
-        (* If env present, then check for incomplete definitions too *)
+        (* If env present, then check for incomplete definitions too;
+           returns both normal variables and row variables*)
+val free_non_row_variables_of_list: type_expr list -> type_expr list
+        (* gets only non-row variables *)
```
Comment (ccasin) [inline typing/ctype.mli:435]: Probably both or neither of these should have "of_list" in the name.  Neither seems better to me, but either is fine.

File: tools/eqparsetree.ml [hunk 19]
```diff
@@ -179,15 +187,18 @@ and eq_core_type_desc :
...
   | (Ptyp_package a0, Ptyp_package b0) -> eq_package_type (a0, b0)
+  | (Ptyp_layout (a0, a1), Ptyp_layout (b0, b1)) ->
+       (eq_core_type (a0, b0)) && (Asttypes.eq_layout_annotation (a1, b1))
   | (_, _) -> false
 and eq_core_type : (core_type * core_type) -> 'result =
   fun
     ({ ptyp_desc = a0; ptyp_loc = a1 },
      { ptyp_desc = b0; ptyp_loc = b1 })
-    -> (eq_core_type_desc (a0, b0)) && (Location.eq_t (a1, b1))
+  -> (eq_core_type_desc (a0, b0)) && (Location.eq_t (a1, b1))
```
Comment (ccasin) [inline tools/eqparsetree.ml]: (spurious whitespace change)

File: typing/ctype.ml [hunk 20]
```diff
@@ -152,7 +152,8 @@ let global_level = s_ref 1
 let saved_level = s_ref []
 
 type levels =
-    { current_level: int; nongen_level: int; global_level: int;
+    { current_level: int; nongen_level: int;
+      global_level: int; (* level assigned to a fresh 'a in user code *)
```
Comment (ccasin) [inline typing/ctype.ml:155]: Minor nit, but I think it's better to put this comment before or after the type to reduce the risk of future merge conflicts with upstream.  Not that this type is particularly likely to change.

File: typing/ctype.mli [hunk 21]
```diff
@@ -469,6 +473,10 @@ val get_unboxed_type_representation : Env.t -> type_expr -> type_expr
 (* Cheap upper bound on layout.  Will not expand unboxed types - call
    [type_layout] if that's needed. *)
 val estimate_type_layout : Env.t ->  type_expr -> Type_layout.t
+
+(* Get an upper bound on layout. (Returns [any] if a type isn't in the env or
+   if the unboxed-unroller runs out of fuel. This uses [repr] so you don't
+   have to. *)
```
Comment (ccasin) [inline typing/ctype.mli]: I suspect the comment about "repr" here predates the 4.14 rebase and can be deleted.

I think it's a bit misleading to say this gets an upper bound on the layout.  The idea is that `estimate_type_layout` returns an upper bound, and this function returns the best layout we have.  It's true that there are some corner cases where that layout may be surprising (as in the case of a missing cmi file) but whatever answer we get is the actual layout as far as the type checker is concerned.

And, as a final nit, we don't necessarily return `any` if we run out of fuel.  Some care is taken to return the layout of the "most" expanded version of the type we encountered.

File: typing/printtyp.ml [hunk 22]
```diff
@@ -1760,12 +1866,20 @@ let class_type ppf cty =
...
-  (match tree_of_typexp Type_scheme param with
-    Otyp_var (_, s) -> s
-  | _ -> "?"),
-  if is_Tvar param then Asttypes.(NoVariance, NoInjectivity)
-  else variance
+let tree_of_class_param param var_inj =
+  let variance, injectivity = if is_Tvar param
+                                   then Asttypes.(NoVariance, NoInjectivity)
+                                   else var_inj
+  in
```
Comment (ccasin) [inline typing/printtyp.ml:1872]: Wonky indenting here - probably just put the whole `if` on its own line.

File: typing/printtyped.ml [hunk 23]
```diff
@@ -155,8 +155,20 @@ let arg_label i ppf = function
   | Labelled s -> line i ppf "Labelled \"%s\"\n" s
 ;;
 
+let typevar_layout ?(print_quote = true) ppf (v, l) =
```
Comment (ccasin) [inline typing/printtyped.ml]: Same comment about default arguments as in the other printers.

File: typing/printtyped.ml [hunk 24]
```diff
@@ -196,6 +208,11 @@ let attributes i ppf l =
     Printast.payload (i + 1) ppf a.Parsetree.attr_payload
   ) l
 
+let var_layout ppf (v, l) =
```
Comment (ccasin) [inline typing/printtyped.ml]: Is this function duplicative of `typevar_layout`?

File: typing/typecore.ml [hunk 25]
```diff
@@ -4860,10 +4860,9 @@ and type_expect_
       in
       re { exp with exp_extra =
              (Texp_poly cty, loc, sexp.pexp_attributes) :: exp.exp_extra }
-  | Pexp_newtype({txt=name}, sbody) ->
+  | Pexp_newtype({txt=name}, sbody, lay) ->
       let layout =
-        Type_layout.of_layout_annotation ~default:Type_layout.value
-          (Builtin_attributes.layout sexp.pexp_attributes)
+        Type_layout.of_layout_annotation_opt ~default:Type_layout.value lay
```
Comment (ccasin) [inline typing/typecore.ml]: I'm wondering if we need to keep around some support for the attribute form of layout annotations to support public release code.  But perhaps we'll sort this out when we move to the new extensions mechanism.

File: typing/typetexp.mli [hunk 26]
```diff
@@ -20,8 +20,14 @@ open Types
...
-val make_poly_univars : string list -> poly_univars
-  (* Create a set of univars with given names *)
+val make_poly_univars : string Location.loc list ->
+  Asttypes.layout_annotation Location.loc option list ->
+  poly_univars
+  (* Create a set of univars with given names and layouts,
+     returning the layouts of the created univars. Note
+     that any call of this function *must* call [check_poly_univars]
+     when done e.g. translating the type in which the univars
+     are in scope. *)
```
Comment (ccasin) [inline typing/typetexp.mli]: I'm confused by the remark that this "return[s] the layouts of the created univars".  Surely the type `poly_univars` represents the univars along with their layouts, not just the layouts.

File: typing/typetexp.ml [hunk 27]
```diff
@@ -53,10 +62,20 @@ type error =
...
 
 exception Error of Location.t * Env.t * error
 exception Error_forward of Location.error
 
+type poly_univars = (string * type_expr * layout_info) list
+
+let rec find_poly_univars name = function
+  | [] -> raise Not_found
+  | (n, t, _) :: rest ->
+    if name = n
```
Comment (ccasin) [inline typing/typetexp.ml]: Maybe `String.equal` instead of polymorphic equality.

File: typing/typetexp.ml [hunk 28]
```diff
@@ -29,6 +29,13 @@ exception Already_bound
 type value_loc =
     Fun_arg | Fun_ret | Tuple | Poly_variant | Package_constraint | Object_field
 
+type cannot_quantify_reason =
+  | Unified of type_expr
+  | Univar
+  | Scope_escape
+
+type layout_info = { original_layout : layout; defaulted : bool }
```
Comment (ccasin) [inline typing/typetexp.ml:43]: I was initially confused by the name `original_layout`, because it suggests the layout might change, and of course we check it doesn't.  I understand now that it's chosen in contrast to `inferred_layout` in the inline record below - perhaps a comment to that effect would be helpful.

File: typing/typetexp.ml [hunk 29]
```diff
@@ -44,7 +51,9 @@ type error =
   | Not_a_variant of type_expr
   | Variant_tags of string * string
   | Invalid_variable_name of string
-  | Cannot_quantify of string * type_expr
+  | Cannot_quantify of string * cannot_quantify_reason
+  | Bad_univar_layout of
+      { name : string; layout_info : layout_info; inferred_layout : layout }
```
Comment (ccasin) [inline typing/typetexp.ml:62]: Should `Bad_univar_layout` be a `cannot_quantify_reason` instead of its own `error`?

File: typing/typedtree.mli [hunk 30]
```diff
@@ -755,7 +762,7 @@ and constructor_declaration =
     {
      cd_id: Ident.t;
      cd_name: string loc;
-     cd_vars: string loc list;
+     cd_vars: (string * layout_annotation option) list;
```
Comment (ccasin) [inline typing/typedtree.mli]: I guess since this typechecks the `loc` information wasn't used anywhere, but I wonder if it wouldn't be nice to keep and use to give better located errors when something fails in layout checking, instead of indicating the whole quantified type.  (and in `Text_decl` below, as well).  Though I guess if we do that here we'd want to do it in all places you can have quantifiers like this.

**Review by ccasin — COMMENTED** [review]
This looks mostly good.  I think there is one issue, which is to do with `check_poly_univars` and mutually defined types.  For example, this currently checks:

```
type ('a : immediate) t_imm

type s = { f : ('a : value) . 'a -> 'a u }
and 'a u = 'a t_imm
```

I think we want to reject this.  (Or if we want to accept it, `s` should get the type `{f : ('a : immediate). 'a -> 'a u}`, but it gets `value`).

The problem is that both `make_poly_univars` and `check_poly_univars` happen inside the call to `transl_simple_type` in `transl_labels`, and that happens in an environment where we haven't yet discovered that `u`'s parameter must be immediate.  There's a similar problem for gadt constructors.

I don't see a super easy solution, and it's further complicated by the fact that we aren't even doing layout checks when these types are translated (because they are delayed for later).  Three possible but annoying solutions:
1) It might be possible to remember these univar checks and perform them later, after we do all the delayed layout checks (though I'd have to think a little harder about what information ends up in what environment). This seems easiest if possible
2) We could retranslate these types for the purpose of performing this check once we have an appropriate environment and have performed the circularity checks.  This seems like a bad idea for performance reasons.
3) We could figure out what we're going to do about "lower bounds" for layouts, which we already want for efficiency and I think could solve this (if `u`'s parameter gets a lower bound of `value` while translating `s`'s body, then when we could get an error when translating `u`).

I'm happy to think more about this - I really want to get out this review today, so I haven't spent much time on it.

I've put some other comments and design questions throughout, but overall things look good other than this issue.  There are a few things that seem a bit awkward, but which I've mostly held off on commenting on because I expect they will change when we eliminate the modifications to the parse tree and move to the new extension mechanism (e.g., there are lots of places where we're using a pair of lists of vars and layout annotations where it seems cleaner to have a list of pairs).

File: testsuite/tests/typing-unboxed-types/test.ml (hunk 3 above)
Comment (goldfirere) [reply testsuite/tests/typing-unboxed-types/test.ml]: In these recursive cases, for unboxed types, it might indeed be appropriate to include a `: value` annotation. But I don't think it should depend on what the user wrote. Instead, I would want to print out a canonical syntax for the type.

Looking around for prior art in this space, I thought about the `-unboxed-types` option, which automatically unboxes all types where this is possible. If I have

```ocaml
type t = MkT of int
```

and say `ocamlc -i` on the file, I get

```
type t = MkT of int
```

Good. Now if I say `ocamlc -unboxed-types -i`, I get

```
type t = MkT of int [@@unboxed]
```

No change in source code. But ocaml is printing out the canonical declaration each time.

On the other hand, if I say

```ocaml
type t = MkT of int [@@boxed]
```

and compile with `ocamlc -unboxed-types -i`, I get

```
type t = MkT of int
```

This is still my canonical representation, but it's potentially confusing, because if I change my code to have that output, it would mean something different than what I wrote (given the use of the `-unboxed-types` flag). So maybe it's "canonical representation in the absence of flags". In any case, it doesn't take into account the user's precise syntax.

Happy to discuss further!

File: parsing/ast_mapper.mli (hunk 6 above)
Comment (goldfirere) [reply parsing/ast_mapper.mli]: I agree. Let's just see how it plays out.

File: parsing/ast_helper.mli (hunk 9 above)
Comment (goldfirere) [reply parsing/ast_helper.mli]: In many places, I would use a list of pairs, or rather a custom datatype. But I know I'm going to have to move the layouts out into attributes in the end, and so I thought the current structure would make that refactor easier. I might well be wrong, but I'm with you in preferring a list of pairs (at least). I should have commented this in the PR.

File: parsing/pprintast.mli (hunk 10 above)
Comment (goldfirere) [reply parsing/pprintast.mli]: I'm not as tempted to do that, specifically. Parsetree and typedtree have their own pretty-printers, after all. And I'm not all that excited to have the parser depend on the type checker, even if doing so doesn't create actual circular dependencies.

That said, I agree that things could potentially get out of sync. Maybe we should have `Type_layout.Const.t` just be `Asttypes.layout_annotation` and then have a `inspect_layout : layout -> (Const.t -> 'a) -> (sort option ref -> 'a) -> 'a`. Then we could avoid duplication. I do think the `Const.t = layout_annotation` is a good idea (and it should remain true as we grow), but I'm less convinced about `inspect_layout`. That said, if you think `inspect_layout` is a good idea, let me know and I can make the refactor.

File: parsing/printast.ml (hunk 11 above)
Comment (goldfirere) [reply parsing/printast.ml]: Agreed about the default parameter; I removed. And I put quotes in.

File: parsing/pprintast.ml (hunk 13 above)
Comment (goldfirere) [reply parsing/pprintast.ml:1362]: Comment added. Just follow the information flow.

File: stdlib/camlinternalAtomic.mli (hunk 14 above)
Comment (goldfirere) [reply stdlib/camlinternalAtomic.mli]: Weird. I have no memory of this. I've reverted.

File: parsing/parser.mly (hunk 15 above)
Comment (goldfirere) [reply parsing/parser.mly:776]: Agreed about delicate design here. I'd be happy for something different than what I've written. We could say that labels require having no space and layouts require having a space, but that's not backward compatible and will make upstreaming harder than it would be otherwise. And it makes a spacebar significant instead of a parenthesis, which is worse. (We could require both whitespace *and* parens. Maybe that's a good idea? I don't know.)

I don't like the keywords idea: we're eventually going to allow abstract layouts (e.g. `sig layout l type t : l ... end`) and so the keywords approach will run out of steam pretty soon.

On balance, my call would be to keep what I've written. But it's more because I don't see a better option than that I like what we have, and so more brainstorming might be fruitful. Maybe add to an agenda for a type-systems meeting? (To be clear: I don't plan on adding this to an agenda myself, but would happily support you if you do.)

Comment (goldfirere) [reply parsing/parser.mly:776]: (Leaving this for you to resolve so that the comment doesn't get lost.)

File: parsing/parser.mly (hunk 16 above)
Comment (goldfirere) [reply parsing/parser.mly:3325]: Bingo.

File: typing/printtyp.ml (hunk 17 above)
Comment (goldfirere) [reply typing/printtyp.ml:87]: Let's talk about this with the team. I see what you're saying, but I'm not yet convinced. The key question, to me: how often is this form of restricted polymorphism actually used? That is, how many functions are polymorphic, but only over immediates? I would expect this to be relatively rare. (But I could very well be wrong.)

I do think your alternative approach is viable. That is, I think most of the places that pretty-print could have an env if they wanted. (It's usually from `report_error`, and so we'd have to add an env to the error exception in e.g. typecore, but that's very doable.) And we can easily make the env be optional and just fall back to my C3 if it's missing. I'll put this in my list of topics to discuss with the team.

File: typing/ctype.mli (hunk 18 above)
Comment (goldfirere) [reply typing/ctype.mli:435]: But one takes a list and the other doesn't. Maybe you were looking at the return type? That is, there really are two differences between the functoins, but these are the only variants we actually need.

File: typing/ctype.ml (hunk 20 above)
Comment (goldfirere) [reply typing/ctype.ml:155]: Fair point. I've taken a different approach of putting each field on its own line.

File: typing/ctype.mli (hunk 21 above)
Comment (goldfirere) [reply typing/ctype.mli]: All good points. Thanks. Left open for you to resolve if you are content.

File: typing/typecore.ml (hunk 25 above)
Comment (goldfirere) [reply typing/typecore.ml]: I think we will need support for the attribute form, for exactly that reason. But the new extensions mechanism handles this nicely.

File: typing/typetexp.ml (hunk 28 above)
Comment (goldfirere) [reply typing/typetexp.ml:43]: Yes, a comment would be good. I've added one. Resolve if you like it.

File: typing/typetexp.ml (hunk 29 above)
Comment (goldfirere) [reply typing/typetexp.ml:62]: It was, originally, but I found the resulting error message -- mentioning "cannot quantify" -- confusing.

File: typing/typedtree.mli (hunk 30 above)
Comment (goldfirere) [reply typing/typedtree.mli]: That's about right. I removed the `loc` to make code reuse easier. Making different decisions about this in different places is annoying. I'm all for adding more locations to report better errors, but I think we should do so separately.

File: testsuite/tests/typing-unboxed-types/test.ml (hunk 3 above)
Comment (ccasin) [reply testsuite/tests/typing-unboxed-types/test.ml]: Sure - I agree canonicity in this sense would be great.  If you have in mind an analysis that will print `[@@value]` only if it's needed, or some overapproximation of that, I'm all for it.

If we have to pick between printing a canonical representation and printing a type that is accurate, I'd certainly choose the latter.  The current behavior for `@@immediate` and `@@immediate64` is to print them just when the user wrote them, and this doesn't seem to cause any issues, despite not being canonical.

File: parsing/pprintast.mli (hunk 10 above)
Comment (ccasin) [reply parsing/pprintast.mli]: I think `Const.t = layout_annotation` is a great idea.  I might prefer a name like `const_layout` in that case.  I'm similarly ambivalent about `inspect_layout`, and don't think the two printers is a big problem.

File: parsing/pprintast.ml (hunk 13 above)
Comment (ccasin) [reply parsing/pprintast.ml:1362]: Ah, right, thanks!

File: typing/ctype.mli (hunk 18 above)
Comment (ccasin) [reply typing/ctype.mli:435]: Sorry, yes - not sure what I was thinking.

File: typing/ctype.mli (hunk 21 above)
Comment (ccasin) [reply typing/ctype.mli]: Looks great, thanks.

File: typing/typetexp.ml (hunk 29 above)
Comment (ccasin) [reply typing/typetexp.ml:62]: Good enough for me, thanks.

File: typing/typedtree.mli (hunk 30 above)
Comment (ccasin) [reply typing/typedtree.mli]: Sounds reasonable.

File: parsing/parser.mly (hunk 15 above)
Comment (ccasin) [reply parsing/parser.mly:776]: Yes, I agree that if we want abstract layouts (and we can't find a clever way to syntactically distinguish them - personally I think all emojis would be good), this is an acceptable compromise.  I think it may be worth adding to the agenda to discuss with the team along with the printing stuff, in case anyone has a clever idea.

File: testsuite/tests/typing-unboxed-types/test.ml (hunk 3 above)
Comment (goldfirere) [reply testsuite/tests/typing-unboxed-types/test.ml]: I forgot to read my own Note when replying. Though canonicity would be great, that's not actually what's implemented. What's implemented is that we print any user-written annotation on an unboxed type except value. So, basically, I think you and I have the same stance here, except for what to do when the layout is `value`. I think e.g. `type loopy = MkL of loopy [@@unboxed]` should get a `value` layout, and then (I believe) the current behavior is correct. I'll add a test about the `loopy` behavior (admitting the possiility of further discussion on design) so we don't forget to address this.

Comment (goldfirere) [reply testsuite/tests/typing-unboxed-types/test.ml]: Actually, as I was clarifying the note and the implementation, I've changed my mind, to agree with you. Now, for fully abstract types (i.e. no `=` signs) and for any unboxed types, we're just printing what the user wrote (or not).

File: parsing/parser.mly (hunk 15 above)
Comment (goldfirere) [reply parsing/parser.mly:776]: OK. I've put this on my list of planned items for next week's type-systems meeting. Resolving, knowing we can return later. I've also made https://jira/browse/TANDC-2078 so that we don't forget this.

File: typing/printtyp.ml (hunk 17 above)
Comment (goldfirere) [reply typing/printtyp.ml:87]: I've made https://jira/browse/TANDC-2080 to be sure we don't forget this. Resolving for now.

File: parsing/pprintast.mli (hunk 10 above)
Comment (goldfirere) [reply parsing/pprintast.mli]: This is now done in https://github.com/ocaml-flambda/ocaml-jst/pull/99. Resolving this conversation here.

**Comment by goldfirere** [comment]
OK. I think I've resolved everything here. OK to merge?

**Comment by goldfirere** [comment]
Small update: it was decided not to merge this for now, because we're hoping to release a version of unboxed types without annotation support.

**Comment by goldfirere** [comment]
Superseded by https://github.com/ocaml-flambda/flambda-backend/pull/1417
